### ISSUE TITLE
[Feature Request]: Self-RAG support (NEW TECHNIQUE)

### ISSUE THREAD

**Issue by casper-hansen** [run-llama/llama_index#8502]
### Feature Description

Self-RAG is a new open-source technique (MIT license) that implements:
1. **Adaptive retrieval via retrieval tokens:** allows you to fine-tune LLMs to output `[Retrieval]` tokens mid-generation to indicate when to perform retrieval. It has been empirically shown to improve open-source models to match ChatGPT level of performance in RAG tasks.
2. **Critique tokens:**
    - w_rel (default 1.0): This variable controls the emphasis on relevance during beam search.
    - w_sup (default 1.0): This variable controls the emphasis on support from the document during beam search.
    - w_use (default 0.5): This variable controls the emphasis on overall quality during beam search.

Requirements:
- make it compatible with vLLM for inference such that any fine-tuned Self-RAG model can be deployed in a framework with PagedAttention implemented.

Website: https://selfrag.github.io/
Model: https://huggingface.co/selfrag/selfrag_llama2_7b
Code: https://github.com/AkariAsai/self-rag

![image](https://github.com/run-llama/llama_index/assets/27340033/9911d86d-e44e-49b1-95ef-87f3b9a9e0e6)
![image](https://github.com/run-llama/llama_index/assets/27340033/b262766c-2318-431c-9714-a0738ad66fc4)

**Comment by Disiok** [comment]
Great suggestion! Would love to have a contribution on this!

**Comment by casper-hansen** [comment]
> Great suggestion! Would love to have a contribution on this!

Agreed, I would love it if either the original authors @AkariAsai or a LlamaIndex maintainer could implement it. It would enable better RAG applications.

**Comment by tanmaylaud** [comment]
Is this implemented ? Or any progress ?
